module/executiondatasync/requester/fulfiller.go

```diff
@@ -0,0 +1,199 @@
+
+	logger  zerolog.Logger
+	metrics module.ExecutionDataRequesterMetrics
+
+	notifier *notifier
+	storage  tracker.Storage
+
+	component.Component
+}
+
+func newFulfiller(
+	fulfilledHeight uint64,
+	notifier *notifier,
+	storage tracker.Storage,
+	logger zerolog.Logger,
+	metrics module.ExecutionDataRequesterMetrics,
+) *fulfiller {
+	jobsIn, jobsOut := util.UnboundedChannel()
+	resultsIn, resultsOut := util.UnboundedChannel()
+
+	f := &fulfiller{
+		jobResults:      make(map[uint64]map[flow.Identifier]*jobResult),
+		sealedResults:   make(map[uint64]flow.Identifier),
+		fulfilled:       make(map[uint64]*jobResult),
+		jobsIn:          jobsIn,
+		jobsOut:         jobsOut,
+		resultsIn:       resultsIn,
+		resultsOut:      resultsOut,
+		fulfilledHeight: fulfilledHeight,
+		sealedHeight:    fulfilledHeight,
+		notifier:        notifier,
+		storage:         storage,
+		logger:          logger.With().Str("subcomponent", "fulfiller").Logger(),
+		metrics:         metrics,
+	}
+
+	cm := component.NewComponentManagerBuilder().
+		AddWorker(f.loop).
+		Build()
+
+	f.Component = cm
+
+	return f
+}
+
+func (f *fulfiller) submitJobResult(j *jobResult) {
+	f.jobsIn <- j
+}
+
+func (f *fulfiller) submitSealedResult(resultID flow.Identifier) {
+	f.resultsIn <- resultID
+}
+
+func (f *fulfiller) fulfill() error {
+loop:
+	if j, ok := f.fulfilled[f.fulfilledHeight+1]; ok {
+		f.fulfilledHeight++
+		delete(f.fulfilled, f.fulfilledHeight)
+		delete(f.sealedResults, f.fulfilledHeight)
+		f.notifier.notify(j.blockHeight, j.executionData)
+
+		goto loop
+	}
+
+	f.logger.Debug().Uint64("height", f.fulfilledHeight).Msg("updating fulfilled height")
+
+	err := f.storage.SetFulfilledHeight(f.fulfilledHeight)
+	if err == nil {
+		f.metrics.FulfilledHeight(f.fulfilledHeight)
+	}
+	return err
+}
+
+func (f *fulfiller) handleJobResult(ctx irrecoverable.SignalerContext, j *jobResult) {
+	if j.blockHeight <= f.fulfilledHeight {
+		return
+	}
+
+	if j.blockHeight <= f.sealedHeight {
+		if j.resultID != f.sealedResults[j.blockHeight] {
+			f.logger.Debug().
+				Str("result_id", j.resultID.String()).
+				Str("block_id", j.executionData.BlockID.String()).
+				Uint64("block_height", j.blockHeight).
+				Msg("dropping job result")
+			f.metrics.ResultDropped()
+			return
+		}
+
+		if j.err != nil {
+			// sealed execution data failed with non-retryable error
+			ctx.Throw(fmt.Errorf("failed to get sealed execution data for block height %d: %w", j.blockHeight, j.err))
+		}
+
+		f.fulfilled[j.blockHeight] = j
+
+		if j.blockHeight == f.fulfilledHeight+1 {
+			if err := f.fulfill(); err != nil {
+				ctx.Throw(fmt.Errorf("failed to fulfill: %w", err))
+			}
+		}
+	} else {
+		jmap, ok := f.jobResults[j.blockHeight]
+		if !ok {
+			jmap = make(map[flow.Identifier]*jobResult)
+			f.jobResults[j.blockHeight] = jmap
+		}
+
+		jmap[j.resultID] = j
+	}
+}
+
+func (f *fulfiller) handleSealedResult(ctx irrecoverable.SignalerContext, sealedResultID flow.Identifier) {
+	f.sealedHeight++
```

Comment:
⚠️  There is a critical undocumented assumption here. For `fulfiller`'s `sealedHeight` state to be correct, sealed results must be delivered exactly once and in the correct order. 

It would be better for this function to accept the block height corresponding to `sealedResultID`, then we can set `f.sealedHeight = max(f.sealedHeight, sealedResultHeight)` and have less rigid constraints on inputs. If we need to keep the assumption about inputs, please note it in the function documentation.